A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,668 @@
+
+```proto
+message WeightedTargetConfig {
+  message Target {
+    uint32 weight = 1;
+    repeated LoadBalancingConfig child_policy = 2;
+  }
+  map<string, Target> targets = 1;
+}
+```
+
+The name (key of the map) will allow the LB policy to be intelligent about
+updating existing children instead of needlessly recreating them when it
+receives an updated config.
+
+The policy will create the child policy for each target. It will then
+distribute picks to the child policies based on the specified weights.
+
+For the purpose of traffic splitting, the child policy will always be
+`cds_experimental`. But note that this policy can be generalized to have
+other LB policies as children.
+
+The way child policies are managed is similar to how the EDS policy manages
+its child policies for localities (in fact, this policy can be reused as a
+child of the EDS policy, but that’s out of the scope of this design). The
+connectivity state of the policy will be the aggregated state of all child
+policies. The WRR algorithm will be random based. When a child policy is
+removed, it will be cached for 15 minutes.
+
+## Routing LB policy
+
+The routing LB policy will have a group of child policies, and their
+corresponding matching criteria. The RPCs will be matched against the
+criteria, and will be distributed to the matched child.
+
+### Child policies
+
+The child policies will be one of
+
+*   `cds_experimental`
+*   `weighted_target_experimental`
+    *   Because this will only have CDS as children, this is essentially
+        `weighted_CDS`
+
+More child policies can be added in the future.
+
+### LB policy config
+
+The `xds_routing_experimental` LB policy will have the following
+configuration:
+
+```proto
+/*
+
+Route List                         Action map
++---------+--------+               +-------+-----------------+--------------+
+| matcher | action |               | name  | policy          | config       |
++------------------+               +----------------------------------------+
+| /s/m1   | wt_1   | +-----------> | wt_1  | weighted_target | <config_1>   |
++------------------+               +----------------------------------------+
+| /s/m2   | cds_1  | +----\ /----> | wt_2  | weighted_target | <config_2>   |
++------------------+       x       +----------------------------------------+
+| /s2/m3  | wt_2   | +----/ \----> | cds_1 | CDS             | <cds_config> |
++------------------+            -> +-------+-----------------+--------------+
+| /s2/m4  | cds_1  | +---------/
++---------+--------+
+
+ */
+
+message XdsRouting {
+  message Action {
+    repeated LoadBalancingConfig child_policy = 1;
+  }
+  // Route represents a routing rule to send an RPC. It contains a matcher and
+  // the name of the action to take. The action with the name can be found in
+  // the action list.
+  //
+  // Route contains the action name instead of embedding the Action, so Action
+  // can be shared and reused by multiple routes.
+  message Route {
```

Comment:
Why have you redefined Envoy's internal protos here rather than referenced them?

Reply:
We internally pass this config around in JSON form, so having a more compact representation is valuable.